independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Decorators.java

```diff
@@ -0,0 +1,101 @@
+            InjectionPointModifier transformer, AnnotationStore store) {
+
+        // Find the delegate injection point
+        List<InjectionPointInfo> delegateInjectionPoints = new LinkedList<>();
+        List<Injection> injections = Injection.forBean(decoratorClass, null, beanDeployment, transformer);
+        for (Injection injection : injections) {
+            for (InjectionPointInfo injectionPoint : injection.injectionPoints) {
+                if (injectionPoint.isDelegate()) {
+                    delegateInjectionPoints.add(injectionPoint);
+                }
+            }
+        }
+        if (delegateInjectionPoints.isEmpty()) {
+            throw new DefinitionException("The decorator " + decoratorClass + " has no @Delegate injection point");
+        } else if (delegateInjectionPoints.size() > 1) {
+            throw new DefinitionException(
+                    "The decorator " + decoratorClass + " has multiple @Delegate injection points: " + delegateInjectionPoints);
+        }
+
+        InjectionPointInfo delegateInjectionPoint = delegateInjectionPoints.get(0);
+
+        Integer priority = 0;
+        boolean priorityDeclared = false;
+        for (AnnotationInstance annotation : store.getAnnotations(decoratorClass)) {
+            if (annotation.name().equals(DotNames.PRIORITY)) {
+                priority = annotation.value().asInt();
+                priorityDeclared = true;
+            }
+        }
+
+        //  The set includes all bean types which are Java interfaces, except for java.io.Serializable
+        Set<Type> types = Types.getClassBeanTypeClosure(decoratorClass, beanDeployment);
+        Set<Type> decoratedTypes = new HashSet<>();
+        for (Type type : types) {
+            if (type.name().equals(DotNames.SERIALIZABLE)) {
+                continue;
+            }
+            ClassInfo clazz = beanDeployment.getBeanArchiveIndex().getClassByName(type.name());
+            if (Modifier.isInterface(clazz.flags())) {
+                decoratedTypes.add(type);
+            }
+        }
+        if (decoratedTypes.isEmpty()) {
+            throw new DefinitionException("The decorator " + decoratorClass + " has no decorated type");
+        }
+
+        // The delegate type of a decorator must implement or extend every decorated type
+        Set<Type> delegateTypes = Types.getDelegateTypeClosure(delegateInjectionPoint, beanDeployment);
+        for (Type decoratedType : decoratedTypes) {
+            if (!delegateTypes.contains(decoratedType)) {
+                throw new DefinitionException("The delegate type " + delegateInjectionPoint.getRequiredType()
+                        + " does not implement the decorated type: " + decoratedType);
+            }
+        }
+
+        if (!priorityDeclared) {
+            LOGGER.info("The decorator " + decoratorClass + " does not declare any @Priority. " +
+                    "It will be assigned a default priority value of 0.");
```

Comment:
I'm not sure. @manovotn WDYT?

Reply:
This scenario is actually ArC trying to make user app work even though in standard CDI it wouldn't fly (the decorator would be ignored). But since there is no beans.xml enablement in our case, this might well be a user error that they want to fix down the line.
I don't think anyone would ever notice this if it were juts debug level...